Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,132 @@
+
+
+## 1. Why
+
+When deploying Prometheus Agent in Kubernetes, three of the biggest users’ concerns are: load distribution, scalability, and security.
+
+DaemonSet deployment solves all these three concerns:
+* Load distribution: Each Prometheus Agent pod will only scrape the targets located on the same node. Even though the targets on some nodes may produce more metrics than other nodes, the load distribution would be reliable enough.
+* Automatic scalability: When new nodes are added to the cluster, new Prometheus Agent pods will be automatically added in the nodes that meet user-defined restrictions (if any).
+* Security: Since the scraped targets are local to the Prometheus Agent pod (on the same node), the scope of security problems is reduced to each node.
+
+DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since the storage requirement for the Agent mode is pretty light compared to a fully functional Prometheus server.
+
+This deployment mode has been implemented in [Google Cloud Managed Service for Prometheus (GMP)](https://github.com/GoogleCloudPlatform/prometheus-engine/), so we have an implementation example to learn from. Also, since this deployment has been tested and proved with GMP’s userbase, we can count on that it can solve a large enough number of use cases.
+
+
+## 2. Pitfalls of the current solution
+
+The current (StatefulSet) deployment brings long the corresponding pitfalls:
+* Load management & scalability: Since one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster, users would need to calculate/estimate the load and scalability of the whole cluster to decide on replicas and sharding strategies. Estimating cluster-wide load and scalability is a much harder task than estimating node-wide load and scalability.
+* Security: Similarly, cluster-wide security is a much bigger problem than node-wide security.
+
+At the moment, the proposed DaemonSet deployment mode doesn’t mean to be a replacement for the current StatefulSet mode. It’s actually a solution for use cases where StatefulSet may not be the best choice. We’ll keep the StatefulSet mode as long as there’s user need for it.
+
+
+## 3. Audience
+
+Users with use cases where:
+* Scraped load is very large or hard to estimate.
+* Scalability is hard to predict.
+* Security is a big concern.
+* They want to collect node system metrics (e.g. kubelet, node exporter).
```

Comment:
Here it sounds like we can't collect kubelet/node exporter metrics with statefulset, which isn't true 🤔

Reply:
Yeah we can do it with `kubernetesSDConfigs` in `ScrapeConfig`? If there's no advantage of solving this use case with DaemonSet, I'll remove it then.

Reply:
We can also create use ServiceMonitor/PodMonitor, just need to adjust labels between services/pods :) 

Yeah, I'd remove this part

Reply:
> Yeah we can do it with `kubernetesSDConfigs` in `ScrapeConfig`?

I meant we can do this in the StatefulSet mode. This is the go-to solution in StatefulSet right?